### DISCUSSION TITLE
How to incrementally hydrate Redux across page fragments during initial load?

### DISCUSSION THREAD

**Discussion by tigt** [reduxjs/redux-toolkit#1820] [Q&A]
I asked @markerikson about this in his Twitter DMs, and he recommended I post it here.

I’m looking for advice, concerns, caveats, needed changes, etc. from y’all’s deep Redux context, not necessarily proof-of-concept code. I found some of that in #1184, but a lot of it is specific to other technologies.

-----------

For an example of the kind of raw immediate pageload speed I’m targeting (and how we’re sending down chunks of HTTP streams), I wrote a post: [_Skeleton screens, but fast_](https://dev.to/tigt/skeleton-screens-but-fast-48f1)

Building off that, the gist of the output HTML stream generation would look like…

```marko
<await(headerApiCall)>
  <@then|headerData|>
    <SiteHead data=headerData /> <!-- this will be a React root -->
  </@then>
</await>

<await(pageContentApiCall)>
  <@then|pageData|>
    <PageRenderer data=pageData /> <!-- this will be a React root -->
  </@then>
</await>

<await(footerApiCall)>
  <@then|footerData|>
    <SiteFoot data=footerData /> <!-- this will be a React root -->
  </@then>
</await>
```

It’s simplified, but hopefully expresses my point. (This might look similar to `<Suspense>`, but [Marko’s `<await>`](https://markojs.com/docs/core-tags/#await) expresses the same functionality differently. The promises for `<@then>` are injected from the top-level to avoid [the “fetch-as-you-render” waterfall problem the React Suspense docs warn against](https://reactjs.org/docs/concurrent-mode-suspense.html#approach-1-fetch-on-render-not-using-suspense).)

The problem: each of those newly-isolated React roots assume a monolithic Redux store is ready before they render — the SSR response was only as fast as the slowest API call. That’s what I want to avoid; once everything in `headerApiCall` resolves, then we should write an HTTP chunk containing the markup and enough Redux initial state to hydrate it with. The `<SiteHead>` should be interactive even if `footerApiCall` takes 10 seconds to finish.

## Prior art on incremental Redux I found

<dl>
<dt><a href="https://medium.com/@paularmstrong/twitter-lite-and-high-performance-react-progressive-web-apps-at-scale-d28a00e780a3#e674">Twitter Lite and High Performance React Progressive Web Apps at Scale § Batch Actions into a Single Dispatch</a>
<dt><a href="https://nicolasgallagher.com/redux-modules-and-code-splitting/">Necolas / Redux modules and code-splitting</a>
<dd>

Twitter Lite’s approach seems closest to what I want, but I’m not quite plugged-in enough to Redux to tell if what they’re talking about is slightly different than what I’m looking for.

<dt><a href="https://stackoverflow.com/questions/32968016/how-to-dynamically-load-reducers-for-code-splitting-in-a-redux-application">How to dynamically load reducers for code splitting in a Redux application?</a>
<dt><a href="https://dev.to/websavi/how-to-code-split-redux-store-to-further-improve-your-apps-performance-4gg9">How to code split Redux § What about reducers?</a>
<dd>

Route-splitting reducers seems to be fairly-understood and widely-practiced by now, but are there any problems or gotchas from doing so?

<dt>I saw that <a href="https://blog.isquaredsoftware.com/presentations/2021-05-state-of-redux-may-2021/?slideIndex=36&stepIndex=0">RTK Query can handle streaming updates for “initial HTTP request”</a>
<dd>How does that work?

</dl>

## Possibly-important details

- The eventual total transmitted JS bundle goal is 20kB, but any improvement over [our current ≥52.2kB setup](https://bundle.js.org/?q=react,react-dom,redux,react-redux,redux-thunk,redux-act) is welcome. For example:
  
  > **\<@markerikson\>** FWIW, if you're only using the React-Redux hooks instead of `connect`, that would shave a few K off its size

- Suspense/Concurrent Mode/`useMutableSource`/etc. are probably “future best practice”, but this is a problem we’re solving for current customers. (We’re not even on React 17 yet 🤷)

- This is going in an A/B test, so the code doesn’t need to avoid hacks… yet.

- Our Redux code doubles as a fascinating archaeological dig. We’ve got `.ts` reducers next to `const SIGN_IN = 'SIGN_IN'`, slices coexisting with `/reducers/`, etc.

  - Our reducers/thunks/other sundry functions implement a _lot_ of business logic, analytics, etc. that would be a huge pain to extract.

  - It also probably contains everything you could think of. For example, a reducer that writes to `localStorage` to work around a SPA architecture problem.

  - We SSR, so the store is JSON-serializable.

- The streaming is called “chunked transfer-encoding” in HTTP/1.1; HTTP/2 does it differently, but the results are the same.

  - The stream must avoid another HTTP request (like how React 18 is doing the weird-not-quite JSON), instead interpolating data into the root `text/html` response.

  - AJAX requests might modify the still-on-server state in the middle of the HTTP request stream (unless that would be too hard to work with, in which case I’ll play ball).

- Replicating the entire store on the client is unnecessary, and arguably undesirable: it would be great to “quarantine” pieces on the server that clients don’t need.

**Comment by markerikson** [comment]
@ephem is the main expert on Redux-related SSR that I know of. There's actually an open PR for React-Redux related to React 18 SSR support, and the last few comments of that may be relevant:

https://github.com/reduxjs/react-redux/pull/1835

**Comment by Ephem** [comment]
Wow, that's an ambitious project, and a well researched post. I can't comment on everything, but I'll try to pick a few things. 😃

First of all, know that what you are getting into is a pretty big and complex problem space and something not only the React team, but other frameworks as well have spent years thinking about (Marko being early in a lot of regards). What might save you is that you are implementing this for a single specific project with known constraints as opposed to trying to solve for the general case. Also, I'd be wary of locking myself too much into a proprietary solution if that makes it significantly harder to merge with the ecosystem later, but I hear ya, having UX-problems here and now carries a lot of weight.

Redux is one big store, and a low level tool, a lot of complexity is going to be up to you to manage here. A likely problem you might run into here is if separate roots rely on the same state in the store. One root hydrates first and changes the state, now the next root and the server will have rendered with different state, causing a hydration mismatch. This _might_ be solvable in the next major Redux version if you are on React >16.8, but even so, I think you'll probably run into other edge cases around this.

What I think you need to do is isolate the state for the different roots (at least any state that ever changes). This could still be done with a single store, but it might be easier to actually use separate stores for each root to guarantee they don't modify each other (but they could reuse parts of the initial state). I realize this probably isn't easy to do because of all that legacy code, but yeah, that's the tricky part. If you use a single store for all roots, the trick is to make each root only select parts of the store the other roots doesn't change, at least for as long as hydration of any root is still ongoing, but this is tricky to enforce, especially over time.

As for the technical details, how to actually solve this?

If you go for the separate stores route, you pass separate `initialState` into them when you create them. If you choose to have a single store, create a custom hydrate-action and before you do `ReactDOM.hydrate()`, dispatch that hydrate action with the _new_ data for that root. How do you determine which data is _new_ compared to already existing in the store you ask? Good question. How do you handle making sure you don't hydrate state that is already in the store, or handle the case if you still do it? Good question. This is the kind of complexity that there are no shortcuts around (at least that I know of).

A totally separate issue here is of course how to handle script loading. How do you know exactly which scripts needs to be loaded before hydrating a specific root? If you solve that, code splitting reducers and injecting them to the store before dispatching the custom hydrate action should be easy (in comparison).

A lot of this builds on the assumption that you want to start hydrating the roots as early as possible on the client, if it's the streaming html you are after but you're fine with hydrating all roots on the client after _everything_ is done and loaded, some problems might get simpler to tackle. In that case, in [Web Vitals](https://web.dev/vitals/) speak, you would get a good Largest Contentful Paint, but would still have a bad First Input Delay.

Just a few random thoughts, not sure if they help. I haven't done this myself, but feel free to ask if you have any specific questions. 😄

(Note that I'm not a RTK Query user, so I'm not sure how that would handle this case, I'm speaking more generally here.)

**Reply by tigt** [reply]
Thank you so much for the detail and consideration!

> What might save you is that you are implementing this for a single specific project with known constraints as opposed to trying to solve for the general case.

Agreed. Immediate customer benefit and $ returns in a project make deviating from the golden path fairly tolerable, usually.

> A likely problem you might run into here is if separate roots rely on the same state in the store. One root hydrates first and changes the state, now the next root and the server will have rendered with different state, causing a hydration mismatch.

I’m almost certain we’ll run into this problem. Could we run currently-accumulated reducers across each store to do a “rolling sync”, or is that steering towards a known problem?

The ability to only bring in small pieces of the store onto the client could also help a lot mitigating this — it wouldn’t avoid the problem, but it would reduce its impact and probability.

> If you use a single store for all roots, the trick is to make each root only select parts of the store the other roots doesn't change, at least for as long as hydration of any root is still ongoing, but this is tricky to enforce, especially over time.

This might not end up a final constraint, in the longer-term. Client-side routing’s tradeoffs become weighted very differently in this use-case — if we already have to hit the network for any reason, concatenating strings on the server + HTML parsing may or may not have as much impact as:

- Downloading and executing a full-power client-side view library
- Data-fetching libraries, state management, caching, etc. on the client, where the connection’s “last mile” is worse than our own Web server’s connection to our API servers over datacenter backbones
- Downloading and executing code for client-side routing, navigation, [accessibility remediation](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/), etc.
- This is further weighted by the e-commerce domain: we _really_ can’t trust the client with a lot of code or business logic, we already have stateful backends for other reasons, and the vast inventory makes it difficult to cache enough product data on the client to meaningfully improve re-access speed.

…that was a lot of words for me to say “especially over time” may be a non-problem for this project. As long as we have some kind of consistency during the initial load, prefetching/bootstrapping SPA code in the background (lazy execution, in Web/Service Workers, etc.) is a lot more tractable and kinder to the user experience than doing it all ASAP during the critical path.

That said, I still think your suggestion about a store for each root is probably the simplest thing that could possibly work, as you later describe:

> How do you determine which data is new compared to already existing in the store you ask? Good question. How do you handle making sure you don't hydrate state that is already in the store, or handle the case if you still do it? Good question. This is the kind of complexity that there are no shortcuts around (at least that I know of).

In comparison, your description of store-per-root sounds a lot less intimidating:

> If you go for the separate stores route, you pass separate `initialState` into them when you create them.

This might be possible with a `Proxy` or getters/setters letting the server code know which parts of the monolithic Redux SSR store a given React root’s components have read or assigned… or does that sound too clever?

> A totally separate issue here is of course how to handle script loading. How do you know exactly which scripts needs to be loaded before hydrating a specific root?

Hmmmm. We already have build machinery doing that per-route… If a “route” is just a sort of specially-marked import that the bundler knows to treat as an entry point, perhaps the same could be done with further granularity by considering each React root on each route an entry point? (“Root-per-route” — thank goodness this is a textual conversation.)

> A lot of this builds on the assumption that you want to start hydrating the roots as early as possible on the client, if it's the streaming html you are after but you're fine with hydrating all roots on the client after _everything_ is done and loaded, some problems might get simpler to tackle.

Our early-painting metrics are pretty good already thanks to SSR, but it’s the “uncanny valley”/late execution that’s proven to drive away our customers.

By using Marko’s `<await>` without any Marko components using `class`/`state` within them, Marko luckily omits its client runtime and can emit the async data at the end of each HTTP chunk “flush”, which means as far as the server code, appending known Redux store data at the end has a promising path forward. By hooking into that API, the ability to tell each React root it’s time to hydrate should be straightforward… I hope.

> (Note that I'm not a RTK Query user, so I'm not sure how that would handle this case, I'm speaking more generally here.)

That’s exactly the sort of help I was looking for! This problem is big enough to span the entire HTTP/HTML/JS/etc. stack, so restricting ourselves to in-library solutions may not be worth it.

**Reply by phryneas** [reply]
Just a slight note from my side: you would probably not go for one store per page, but for slices that are dynamically injected when they are required/loaded for the first time. Redux is not really made for multiple stores and you won't have a lot of benefits from that.

**Reply by Ephem** [reply]
Yeah, that's a good point @phryneas, main message is rather "find a way to isolate dynamic state between the roots".

Another approach could be to design your own "freezed" state I guess. Before every root has hydrated, dispatches are not allowed to change state, instead queueing and turning buttons into loading states etc. Users can't do much then, but at least they could get immediate feedback that their action is being worked on. This is probably not that simple to implement though.

I'll repeat that _if_ you are at React >16.8, or can get there, the next version of Redux could help. It will probably feature something like:

```<Provider store={store} serverState={initialStateFromServer}>```

What this will do is make sure that during hydration, React renders with the state from the server even if that has later changed, right after hydration it will rerender with any changed state. This could let you worry less about isolating state etc.

**Reply by Ephem** [reply]
Oh, and btw, hydration mismatches aren't the end of the world, React recovers just fine from them (see [this codesandbox](https://codesandbox.io/s/react-18-redux-ssr-npltr) for example), so don't let that stop you, especially from experimenting. 😃

**Reply by tigt** [reply]
Noted. I’ll be sure to run these specifics by any Redux fans we have to see if that gets their gears turning.

**Reply by tigt** [reply]
I’m researching this advice from @phryneas:

> _[…]_ you would probably not go for one store per page, but for slices that are dynamically injected when they are required/loaded for the first time. Redux is not really made for multiple stores and you won't have a lot of benefits from that.

Double-checking; this means [the `createSlice()` API](https://redux-toolkit.js.org/api/createSlice)? So for each HTTP chunk, I’m picturing something like…

```html
<script>
Object.assign(
  window.__ACCUMULATED_INITIAL_STATE__,
  JSON.parse('…serialized store fragment…')
)

reduxInstance.combineReducers(createSlice({
  name: 'site-header',
  initialState: window.__ACCUMULATED_INITIAL_STATE__,
  reducers: { …import('/reducers-for-site-header.js') }
}))
</script>
```

…and then `extraReducers` looks like the no-holds-barred extensibility point, so presumably I’m going to get real comfortable with it considering I’m about to do the definition of unforeseen library extension.

(By the way, nice taste with the convenience reexport of `nanoid`.)

**Reply by phryneas** [reply]
You should really not go to accumulate some initial state. It is the **initial** state. It will not have any effect after store is created.
As already stated: dispatch a hydration action and let your reducers respond to that.

Apart from that: not necessarily one slice per sub-page. Just use slices as you want, but use some kind of lazy injection mechanism (RTK doesn't offer one, but there are enough examples around the net).

Let your bundler figure out what needs to be sent to the client when, using `import()` statements.

Also, I think you are overengineering this "capturing all actions and replaying them once the reducers for them are loaded" thing. Export your actions and reducers from the same file and `import()` both of them. That way if the client dispatched an action, you can be sure that the according reducer is loaded at the same time.

**Reply by tigt** [reply]
Excellent, very good to know, thank you. My code sample was sketched without understanding any of the interplay, so I deeply appreciate the explanations of the consequences.

**Comment by phryneas** [comment]
> I saw that RTK Query can handle streaming updates for “initial HTTP request”
    How does that work? 

That is another kind of streaming - making a HTTP request to an api and getting updates via websocket.


You might want to look into what https://github.com/kirill-konshin/next-redux-wrapper is doing - they allow to hydrate SSR content on a per-page basis.

**Reply by phryneas** [reply]
Also note that when using RTK-Query, you should use [`extractRehydrationInfo`](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering) as that will incrementally merge the RTKQ api slice with existing store data and skip data that should not be shared between state and client (component subscriptions etc).

**Reply by tigt** [reply]
> That is another kind of streaming - making a HTTP request to an api and getting updates via websocket.

Ah, thank you. I can write that off, then, since the latency from that approach is undesirable.

For further context, we’re targeting the spottiest connections and lowest-budget Android phones in broad use today — $30 is a common price. WebSockets can do wonderful things, but the extra connection, latency, and battery drain make me leery of using them for something as core as loading any page on the site.

Next.js _might_ be a long-term goal, since React Suspense+SSR still seems to be weirdly gated-off otherwise.

I’m concerned about this golden path’s direction, though. To get the same SSR techniques as Marko, it seems like the recommended solutions have enough dependencies to price us out of [the 150kB speed limit](https://infrequently.org/2017/10/can-you-afford-it-real-world-web-performance-budgets/):

- `react`
- I heard that streaming SSR will require additional React runtime — like 8kB uncompressed? Could grow or shrink before release
- `react-dom`
- `redux`
- `react-redux`
- `@rtk-incubator/rtk-query`
- Next’s own overhead
- `next-redux-wrapper`
- `@reduxjs/toolkit`
- …and those don’t total to 150kB, but we’re also budgeting for the horrible 3<sup>rd</sup> party analytics’ JS that the business won’t let us do without. As a developer, I can’t solve that problem, but I _can_ engineer around it.
- We also need feature/application code somewhere

In that case, maybe this could end up preferring to be a migration _away_ from React. Its incentives and constraints may not match ours in the long-term.

**Reply by markerikson** [reply]
Note that `@rtk-incubator/rtk-query` is gone - that was the early alpha standalone package.  RTK Query is now accessed as a separate entry point in RTK itself - `@reduxjs/toolkit/query/react`.  But yes, that's its own chunk of logic and bundle size.

My general sense here is that you have such highly specific requirements that a "typical" React+Redux setup isn't the right tools for your problem space, but I also have almost no experience here myself.

What do you mean by "React Suspense+SSR is weirdly gated off"?

**Reply by Ephem** [reply]
It's hard to say for sure without knowing more specifics (like are we talking mostly static content website, or highly dynamic webapp?), but yeah, it does sound from those constraints like React _today_ could be a tricky fit.

Could be that in the future Server Components could be a good fit, if a large part of the page was written as such to avoid passing down JS for any part that is not actually dynamic, but that's really hard to say today, and probably wont be easy to apply to a legacy codebase to start with.

I wish I had some great advice here, rewriting everything usually isn't very feasible and is always pretty risky. Two things that helped us improve performance a lot at work was to start lazy loading reducers, and start lazy loading parts of the site below the fold. We still server render everything for bots, but for users, we only server render the top part of the page until they scroll close to the lazy content (important to use intersectionObserver for googlebot to work). This and other things let us incrementally improve performance a lot over time.

Targeting the kind of low end devices and poor network conditions you are talking about, the tradeoffs might be different though and it might take such a long time to load and render the lazy content that it ends up being a poor experience anyway.

Update: (Maybe you are already doing parts of this/have evaluated it/it wont cut it for you, just thought I'd share our journey)

**Reply by tigt** [reply]
> My general sense here is that you have such highly specific requirements that a "typical" React+Redux setup isn't the right tools for your problem space, but I also have almost no experience here myself.

Yep, agreed. We started with “industry standards” in the past for what seemed like good reasons, but it appears that if we want to exceed the industry’s standard speed, we need to be more unconventional.

> What do you mean by "React Suspense+SSR is weirdly gated off"?

I could have worded that better. I mean how it’s strongly recommended you use a framework implementing Server Components, or write your own, instead of treating them as a feature React itself handles for you.

**Reply by tigt** [reply]
> It's hard to say for sure without knowing more specifics (like are we talking mostly static content website, or highly dynamic webapp?)

Large inventory e-commerce. Product lists and pages are mostly static with islands of user-specific state (is it in cart, applied discounts, etc.), but there’s enough products that Static Site Generators/Next.js-style static generation exhibit pathological performance in the build/deploy steps.

> rewriting everything usually isn't very feasible and is always pretty risky

That was my first try; I wrote a Marko-only prototype that hits the aggressive 20kB goal with Brotli compression. We know the end state is possible, but for business concerns, a gradual migration path is important. I guess that’s why I’m asking Redux tool maintainers for forbidden Redux techniques 😅 

> Targeting the kind of low end devices and poor network conditions you are talking about, the tradeoffs might be different though and it might take such a long time to load and render the lazy content that it ends up being a poor experience anyway.

You hit the nail on the head. Marko’s HTML streaming and ability to automatically only send render code/hydration data for component roots with `state` made a lot of the performance workarounds for our current site hurt more than they help — if half the page is swathes of static `<a href=…><img alt=… src=…></a>`, React is buying us nothing — it’s going to only ever run those components mount/unmount lifecycles, but the user still downloads everything else for them.

**Reply by markerikson** [reply]
That's because at the moment, React absolutely does _not_ "handle that for you" :)

Right now React is "just" a set of primitives.  SSR support is (to my knowledge) mostly just `ReactDOMServer.renderToString/NodeStream()`, and Suspense does not yet work in SSR.

React 18 apparently adds support for Suspense in SSR scenarios, and if you read the SSR-related posts in the React 18 working group it's clear that they are working on adding more in-depth SSR support for complex use cases.  Server Components are _related_ to that but not the only part.  Also, Server Components are currently at the "prototype tech demo" level and there's no full solution available, which is why Shopify's "Hydrogen" framework that uses RSCs has to build up a lot of functionality from scratch around them.

**Reply by Ephem** [reply]
> if half the page is swathes of static `<a href=…><img alt=… src=…></a>`, React is buying us nothing — it’s going to only ever run those components mount/unmount lifecycles, but the user still downloads everything else for them.

I feel this every day. 😁 I mean, I guess what it buys is being able to render the next page by only fetching data, the JS is already cached. First visit bad vs better subsequent navigations/vists etc but yeah, not always a great tradeoff. Server Components will hopefully give us a tool where we can tweak this one to fit each specific case (low JS -> heavy markup navigations vs high JS -> low markup navigations), but it's still early and yeah, the investment will probably be high, especially implementing it yourself. 😄
